### PR TITLE
상품 구현

### PR DESCRIPTION
**Pull request by rkdmin**
### 변경사항

**AS-IS**
상품 등록 구현중 
**TO-BE**
상품 등록 구현 완료
### 테스트 
 
- [ ] 테스트 코드
- [ ] API 테스트

### CODE REVIEW

**Comment by rkdmin** [comment]
|File|Coverage [5.67%]|:x:|
|:-|:-:|:-:|
|[ColorType.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Ftype%2FColorType.java)|100%|:green_apple:|
|[SizeType.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Ftype%2FSizeType.java)|100%|:green_apple:|
|[ItemInfoService.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fservice%2FItemInfoService.java)|38.46%|:x:|
|[ItemController.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fcontroller%2FItemController.java)|28.57%|:x:|
|[CategoryController.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fcontroller%2FCategoryController.java)|21.74%|:x:|
|[BrandController.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fcontroller%2FBrandController.java)|20%|:x:|
|[ModelController.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fcontroller%2FModelController.java)|19.61%|:x:|
|[SellerController.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fcontroller%2FSellerController.java)|18.87%|:x:|
|[ItemService.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fservice%2FItemService.java)|14.73%|:x:|
|[CategoryService.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fservice%2FCategoryService.java)|11.24%|:x:|
|[MemberService.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fservice%2FMemberService.java)|10.33%|:x:|
|[ModelService.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fservice%2FModelService.java)|8.4%|:x:|
|[SellerService.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fservice%2FSellerService.java)|7.94%|:x:|
|[BrandService.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fservice%2FBrandService.java)|7.87%|:x:|
|[GlobalExceptionHandler.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fexception%2FGlobalExceptionHandler.java)|7.29%|:x:|
|[Category.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fentity%2FCategory.java)|3.23%|:x:|
|[ItemInfo.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fentity%2FItemInfo.java)|3.16%|:x:|
|[Brand.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fentity%2FBrand.java)|1.86%|:x:|
|[Model.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fentity%2FModel.java)|1.4%|:x:|
|[Seller.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fentity%2FSeller.java)|1.24%|:x:|
|[Item.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fentity%2FItem.java)|0.99%|:x:|
|[BrandException.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fexception%2FBrandException.java)|0%|:x:|
|[CategoryException.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fexception%2FCategoryException.java)|0%|:x:|
|[ItemException.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fexception%2FItemException.java)|0%|:x:|
|[ModelException.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fexception%2FModelException.java)|0%|:x:|
|[SellerException.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fexception%2FSellerException.java)|0%|:x:|
|[RegCategory.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fmodel%2FRegCategory.java)|0%|:x:|
|[RegSeller.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fmodel%2FRegSeller.java)|0%|:x:|
|[UdtCategory.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fmodel%2FUdtCategory.java)|0%|:x:|
|[RegModel.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fmodel%2FRegModel.java)|0%|:x:|
|[UdtModel.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fmodel%2FUdtModel.java)|0%|:x:|
|[UdtSeller.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fmodel%2FUdtSeller.java)|0%|:x:|
|[RegBrand.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fmodel%2FRegBrand.java)|0%|:x:|
|[RegItem.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fmodel%2FRegItem.java)|0%|:x:|
|[UdtBrand.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fmodel%2FUdtBrand.java)|0%|:x:|
|[RegItemInfo.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Fmodel%2FRegItemInfo.java)|0%|:x:|
|[ErrorCode.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Ftype%2FErrorCode.java)|0%|:x:|
|[OrderStatus.java](https://github.com/rkdmin/fitMe/blob/11acec8f80c851a732b179540259698b49fe8be8/src%2Fmain%2Fjava%2Fcom%2Fzerobase%2Ffitme%2Ftype%2FOrderStatus.java)|0%|:x:|

|Total Project Coverage|7.5%|:x:|
|:-|:-:|:-:|

File: src/main/java/com/zerobase/fitme/controller/BrandController.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    @PreAuthorize("hasRole('ADMIN')") // 관리자만 접속 가능
+    @PostMapping("/register")
+    public String register(@RequestBody @Valid RegBrand.Request request, BindingResult bindingResult){
+        // @valid 발생
+        validation(bindingResult);
+
+        brandService.register(request);
+
+        return "브랜드 등록이 완료되었습니다.";
```
Comment (KimHyoJin) [inline src/main/java/com/zerobase/fitme/controller/BrandController.java:41]: 응답형식이 String형태로 내려가는 것보단 성공여부를 담을 수 있으면 좋을 것 같아요. API응답을 프론트엔드쪽에서 파싱해서 사용하기때문에 일정한 형식에 정해진 형태로 내려가는 것이 좋습니다. 성공여부를 string으로 내려준다면 이를 알기가 좀 어려울 것 같아요

File: src/main/java/com/zerobase/fitme/controller/BrandController.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    @PreAuthorize("hasRole('ADMIN')") // 관리자만 접속 가능
+    @DeleteMapping("/delete/{id}")
+    public String delete(@PathVariable Long id){
+        brandService.delete(id);
+        return "브랜드 삭제가 완료되었습니다.";
+    }
+
+    private static void validation(BindingResult bindingResult) {
```
Comment (KimHyoJin) [inline src/main/java/com/zerobase/fitme/controller/BrandController.java:64]: 요런 validation기능은 AOP를 사용하시면 좋을 것 같아요. 좀더 깔끔하게 코드를 짤수 있고 역할별로 구분을 지을 수 있습니다

File: src/main/java/com/zerobase/fitme/entity/Category.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+@AllArgsConstructor
+@Builder
+@Entity
+public class Category{
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    private String categoryName;
+    private boolean usingYn;
```
Comment (KimHyoJin) [inline src/main/java/com/zerobase/fitme/entity/Category.java:27]: 사용여부를 나타내는 필드라면 using 정도로 나타내도 좋을 것 같아요. 필드자체가 boolean이기때문에 Yn이라는 postfix를 붙여주지 않아도 충분히 의미전달이 될 것 같습니다

File: src/main/java/com/zerobase/fitme/service/BrandService.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+@RequiredArgsConstructor
+@Transactional
+public class BrandService {
+    private final BrandRepository brandRepository;
+
+    /**
+     * 브랜드 등록
+     * @param request
+     */
+    public void register(RegBrand.Request request) {
```
Comment (KimHyoJin) [inline src/main/java/com/zerobase/fitme/service/BrandService.java:32]: request를 그대로 받아와서 사용하는 것보단 controller에서 사용하는 dto와 service에서 사용하는 dto를 분리해주는 것이 좀더 명확할 것 같아요. 혹은 RegBrand.Request의 필드를 각각 param으로 넘겨줘도 좋을 것 같습니다

File: src/main/java/com/zerobase/fitme/service/SellerService.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    public void register(Seller seller) {
+        sellerRepository.save(seller);
+    }
+
+    /**
+     * 판매자 조회
+     * @return
+     */
+    public List<Seller> read() {
+        return sellerRepository.findAll();
```
Comment (KimHyoJin) [inline src/main/java/com/zerobase/fitme/service/SellerService.java:57]: findAll을 사용하실땐 전체를 다 가지고오면 DB에 부하가 있을 수 있으니 paging처리를 해주시면 좋을 것 같아요

File: src/main/java/com/zerobase/fitme/service/ItemService.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+                .regDt(LocalDateTime.now())
+                .itemInfo(null)// enum 이슈..
+                .brand(brand)
+                .seller(seller)
+                .model(model)
+                .build()
+        );
+    }
+
+    private Long calculate(Long price, Long saleRate) {
```
Comment (KimHyoJin) [inline src/main/java/com/zerobase/fitme/service/ItemService.java:67]: 어떤 것을 계산하는 함수인지가 조금 불명확한것 같아요. 메소드명에 할인율을 계산한다는 역할이 표현되면 좋을 것 같습니다

File: src/main/java/com/zerobase/fitme/service/CategoryService.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+     * @param request
+     */
+    public void register(RegCategory.Request request) {
+        if(categoryRepository.findByCategoryName(request.getCategoryName()).isPresent()){
+            throw new CategoryException(ALREADY_EXIST_CATEGORY_NAME);
+        }
+
+        categoryRepository.save(Category.builder()
+                .categoryName(request.getCategoryName())
+                .usingYn(false)// 아직 등록 안함
```
Comment (KimHyoJin) [inline src/main/java/com/zerobase/fitme/service/CategoryService.java:37]: 등록하는 함수인데 usingYn은 false로 처리되는 이유가 있을까요? 플로우가 어떻게 되나요?

**Review by KimHyoJin — APPROVED** [review]
안녕하세요. 밍밍멘토입니다. 간단히 몇가지 코멘트 드렸어요! PR크기가 조금 큰데 다음에는 사이즈를 조금 줄여주시면 보다 자세한 리뷰가 가능할 것 같습니다.
슬랙으로 enum list가 entity에 들어가지 않는다고 말씀주셨는데,  list를 저장할 수 있는 column 타입이 mysql에 뭐가 있을지를 한번 찾아보시면 좋을 것 같아요. JPA라는 것은 결국 객체지향으로 쓰여진 코드를 flat한 데이터베이스의 테이블로 변환시켜주는 orm이기때문에 사용하시는 데이터베이스에 존재하는 column type을 사용하셔야합니다. 고민해보시고 조금더 도움이 필요하시면 슬랙으로 알려주세요!

File: src/main/java/com/zerobase/fitme/service/CategoryService.java (hunk 7 above)
Comment (rkdmin) [reply src/main/java/com/zerobase/fitme/service/CategoryService.java:37]: 리뷰 감사합니다! 우선은 생각했던게 카테고리를 일단 만들고 사용여부가 false면 상품 등록할 때 카테고리가 false이기때문에 read가 안되게 설정하려고 usingYn을 넣었습니다. 카테고리를 사용하고 싶으면 수정을 통해 true로 바꾸면 매니저 또는 판매자가 상품 등록할 때 그 카테고리를 사용하게 하는 플로우 입니다!

File: src/main/java/com/zerobase/fitme/controller/BrandController.java (hunk 1 above)
Comment (rkdmin) [reply src/main/java/com/zerobase/fitme/controller/BrandController.java:41]: ResponseEntity<> 를 사용해서 성공코드 200번과 성공 메세지를 보내는 것이 혹시 괜찮을 방법일까요?

File: src/main/java/com/zerobase/fitme/service/BrandService.java (hunk 4 above)
Comment (rkdmin) [reply src/main/java/com/zerobase/fitme/service/BrandService.java:32]: controller에 쓰는 request와 sevice에서 사용하는 객체 필드가 같아서 전해줬는데 그래도 분리해주는 것이 좋을까요?

**Pull request merged**